db/src/schema.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+use edn::types::Value;
+use errors::*;
+use types::{Attribute, Entid, IdentMap, Schema, SchemaMap, ValueType};
+use values;
+
+impl Schema {
+    pub fn get_ident(&self, x: &Entid) -> Option<&String> {
+        self.entid_map.get(x)
+    }
+
+    pub fn get_entid(&self, x: &String) -> Option<&Entid> {
+        self.ident_map.get(x)
+    }
+
+    pub fn attribute_for_entid(&self, x: &Entid) -> Option<&Attribute> {
+        self.schema_map.get(x)
+    }
+
+    /// Turn Value([[IDENT ATTR VALUE] ...]) into a Mentat `Schema`.
+    pub fn from_ident_map_and_assertions(ident_map: IdentMap, assertions: &Value) -> Result<Schema> {
+        // Convert Value([[IDENT ATTR VALUE] ...]) to vec![(IDENT.to_string(), ATTR.to_string(), VALUE), ...].
+        let triples: Vec<(String, String, &Value)> = match *assertions {
+            Value::Vector(ref datoms) => {
+                datoms.into_iter().map(|datom| {
+                    match datom {
+                        &Value::Vector(ref values) => {
+                            let mut i = values.iter();
+                            match (i.next(), i.next(), i.next(), i.next(), i.next()) {
+                                (Some(add), Some(&Value::NamespacedKeyword(ref ident)), Some(&Value::NamespacedKeyword(ref attr)), Some(value), None) if *add == *values::DB_ADD =>
+                                    Ok((ident.to_string(), attr.to_string(), value)),
+                                _ => Err(ErrorKind::BadSchemaAssertion(format!("Expected [[:db/add IDENT ATTR VALUE] ...], got: {:?}", datom)))
+                            }
+                        },
+                        _ => Err(ErrorKind::BadSchemaAssertion(format!("Expected [[...] ...], got: {:?}", datom)))
+                    }
+                }).collect()
+            },
+            _ => Err(ErrorKind::BadSchemaAssertion(format!("Expected [...], got: {:?}", assertions)))
+        }?;
+
+        let mut schema_map = SchemaMap::new();
+        for (ident, attr, value) in triples {
+            let entid: &i64 = ident_map.get(&ident).ok_or(ErrorKind::BadSchemaAssertion(format!("Could not get ")))?;
+            let attributes = schema_map.entry(*entid).or_insert(Attribute::default());
+
+            // Yes, this is pretty bonkers.  Suggestions appreciated.
```

Comment:
Could you rewrite this as another `match`? Untested, but something like

```rust
match attr.as_str() {
    ":db/valueType" => {
        attributes.value_type = match *value {
            *values::DB_TYPE_REF => ValueType::Ref,
            *values::DB_TYPE_BOOLEAN => ValueType::Boolean,
            ...
        };
    },
```

It's not substantively different but is nicer to look at.